sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -68,7 +68,9 @@ Providing a generic group abstraction encapsulating the shared functionality of
 
 === Attributes
 
-The +[[intel::reqd_sub_group_size(n)]]+ attribute indicates that the kernel must be compiled and executed with a sub-group of size _n_.  The value of _n_ must be a compile-time integral constant expression.  The value of _n_ must be set to a sub-group size that is both supported by the device and compatible with all language features used by the kernel, or device compilation will fail.  The set of valid sub-group sizes can be queried as described below.
+The +[[intel::reqd_sub_group_size(S)]]+ attribute indicates that the kernel must be compiled and executed with a specific sub-group size.  The value of _S_ must be a compile-time integral constant expression or one of the following keywords: +auto+, +core+.  If the value of _S_ is an integer, the kernel should only be submitted to a device that supports that sub-group size (as reported by +info::device::sub_group_sizes+).  If the kernel is submitted to a device that does not support the requested sub-group size, or a device on which the requested sub-group size is incompatible with any language features used by the kernel, the implementation must throw a synchronous exception with the `errc::feature_not_supported` error code from the kernel invocation command.  If the value of _S_ is +auto+, the implementation is free to select any of the valid sub-group sizes associated with the device to which the kernel is submitted; the manner in which the sub-group size is selected is implementation-defined.  If the value of _S_ is +core+, the implementation will select the device's core sub-group size (as reported by the +info::device::core_sub_group_size+ query) for all kernels with this attribute.
```

Comment:
What happens for:
```
constexpr int core = 4;
SYCL_EXTERNAL void f() [[intel::reqd_sub_group_size(core)]];
```
Does it use 4 or does it use `core`? We could avoid this possible ambiguity and the need to treat those special in the FE by making this `intel::core`. That allows us to define in the header  
```
namespace intel { constexpr int core = -1; }
```
and the FE doesn't need to have special casing for this.

Reply:
That's a good point, I hadn't considered that "core" might be used elsewhere as a name.  I don't think `INTEL::core` is much better -- `INTEL::reqd_sub_group_size(INTEL::core)` is pretty long, and as you point out in https://github.com/intel/llvm/pull/2452#discussion_r488198433, we might end up with values like `INTEL::core` occurring in multiple contexts.

Ignoring the open naming issue: Do you think `INTEL::req_core_sub_group_size` would be better?  It's 9 characters shorter and unambiguous.

Reply:
I wasn't thinking that the attribute parameter could be a `constexpr` variable from the application.  I was assuming it had to be a literal number or one of the special tokens "core" or "auto".  Therefore, "core" used in the attribute parameter would be unrelated to any variable named "core".

Allowing a `constexpr` variable (or expression) does seem to introduce other problems.  We could define a variable named `core` in our headers as @rolandschulz suggests, but I'm not sure `auto` is allowed as a variable name.  We could avoid those problems by adding three attributes as @Pennycook suggests: `intel::reqd_sub_group_size(S)`, `intel::reqd_core_sub_group_size`, and `intel::reqd_auto_sub_group_size`.

Or, we could specify that `reqd_sub_group_size()` must take a literal number or one of the special tokens "core" or "auto".  I admit that's less flexible, though.

Reply:
The attributes have been split in 7ccb2eb2ca405984ee69ea9ad928e7310ad056e3.  I've tried to make it clear that `[[sub_group_size(S)]]` requires `S` to be a compile-time constant integer expression, whereas `[[named_sub_group_size(NAME)]]` requires `NAME` to be a special token from a known list.